### PR TITLE
Df transport asm

### PR DESCRIPTION
**Pull request by dflanderka**
- new evaluation of 'assembly_xyz' methods in TransportDG

### CODE REVIEW

**Comment by dflanderka** [comment]
Uncomment of line 291 in assembly_dg.hh causes errors:

1) Needs skip boundary Fields.
2) Fields conc and subdomain cause following error.
    There is uninitialized object FunctionSpace called in FiniteElement::n_components probably.

```
terminate called after throwing an instance of 'boost::exception_detail::clone_impl<ExcSignal>'
  what():  --------------------------------------------------------
Program Error: [ Signal 11 (Segmentation fault) received! ]

** Diagnosting info **
/mnt/c/Users/t4260/vyvoj/flow123d/flow123d/src/system/application_base.cc(59): Throw in function PetscErrorCode petsc_signal_handler(int, void*)
Dynamic exception type: boost::exception_detail::clone_impl<ExcSignal>
[EI_Signal_TAG*] = 11
[EI_SignalName_TAG*] = Segmentation fault

** Stacktrace **
  0  petsc_signal_handler(int, void*)
  1  ) [0x1ef51ed]
  2
  3  FunctionSpace::n_components() const
  4  FiniteElement<1u>::n_components() const
  5  FEValues<3u>::FEInternalData* FEValues<3u>::init_fe_data<1u>(FiniteElement<1u> const&, Quadrature const&)
  6  void FEValues<3u>::initialize<1u>(Quadrature&, FiniteElement<1u>&, UpdateFlags)
  7  FieldFE<3, FieldValue_<1, 1, double> >::cache_update(FieldValueCache<double>&, ElementCacheMap&, unsigned int)
  8  Field<3, FieldValue_<1, 1, double> >::cache_update(ElementCacheMap&)
  9  MultiField<3, FieldValue_<1, 1, double> >::cache_update(ElementCacheMap&)
 10  FieldSet::cache_update(ElementCacheMap&)
 11  GenericAssembly<TransportDG<ConcentrationTransportModel>::InitConditionAssemblyDim>::add_integrals_of_computing_step(DHCellAccessor)
 12  GenericAssembly<TransportDG<ConcentrationTransportModel>::InitConditionAssemblyDim>::assemble(std::shared_ptr<DOFHandlerMultiDim>)
 13  TransportDG<ConcentrationTransportModel>::set_initial_condition()
 14  TransportDG<ConcentrationTransportModel>::zero_time_step()
 15  TransportOperatorSplitting::zero_time_step()
 16  HC_ExplicitSequential::advection_process_step(HC_ExplicitSequential::AdvectionData&)
 17  HC_ExplicitSequential::run_simulation()
 18  Application::run()
```

**Comment by jbrezmorf** [comment]
> Uncomment of line 291 in assembly_dg.hh causes errors:
> 
>     1. Needs skip boundary Fields.
> 
>     2. Fields 'conc' and 'subdomain' cause following error.
>        There is uninitialized object FunctionSpace called in FiniteElement::n_components probably.
> 
> 
> ```
> terminate called after throwing an instance of 'boost::exception_detail::clone_impl<ExcSignal>'
>   what():  --------------------------------------------------------
> Program Error: [ Signal 11 (Segmentation fault) received! ]
> 
> ** Diagnosting info **
> /mnt/c/Users/t4260/vyvoj/flow123d/flow123d/src/system/application_base.cc(59): Throw in function PetscErrorCode petsc_signal_handler(int, void*)
> Dynamic exception type: boost::exception_detail::clone_impl<ExcSignal>
> [EI_Signal_TAG*] = 11
> [EI_SignalName_TAG*] = Segmentation fault
> 
> ** Stacktrace **
>   0  petsc_signal_handler(int, void*)
>   1  ) [0x1ef51ed]
>   2
>   3  FunctionSpace::n_components() const
>   4  FiniteElement<1u>::n_components() const
>   5  FEValues<3u>::FEInternalData* FEValues<3u>::init_fe_data<1u>(FiniteElement<1u> const&, Quadrature const&)
>   6  void FEValues<3u>::initialize<1u>(Quadrature&, FiniteElement<1u>&, UpdateFlags)
>   7  FieldFE<3, FieldValue_<1, 1, double> >::cache_update(FieldValueCache<double>&, ElementCacheMap&, unsigned int)
>   8  Field<3, FieldValue_<1, 1, double> >::cache_update(ElementCacheMap&)
>   9  MultiField<3, FieldValue_<1, 1, double> >::cache_update(ElementCacheMap&)
>  10  FieldSet::cache_update(ElementCacheMap&)
>  11  GenericAssembly<TransportDG<ConcentrationTransportModel>::InitConditionAssemblyDim>::add_integrals_of_computing_step(DHCellAccessor)
>  12  GenericAssembly<TransportDG<ConcentrationTransportModel>::InitConditionAssemblyDim>::assemble(std::shared_ptr<DOFHandlerMultiDim>)
>  13  TransportDG<ConcentrationTransportModel>::set_initial_condition()
>  14  TransportDG<ConcentrationTransportModel>::zero_time_step()
>  15  TransportOperatorSplitting::zero_time_step()
>  16  HC_ExplicitSequential::advection_process_step(HC_ExplicitSequential::AdvectionData&)
>  17  HC_ExplicitSequential::run_simulation()
>  18  Application::run()
> ```

No proste se to musi odladit. Nic obecneho me k tomu nenapada. Asi je to nejaka kombiace, kterou by to pripadne chtelo otestovat v unit testu (urcite pripad boundary fields). Tam vime, ze to nebude fungovat pro boundary FieldFE. Ta druha chyba muzebyt napr. zpusobena referenci na nejay temporary objekt. 

Pouzit FEValues::initialize v cache_update by nemelo byt. Initialize by se melo volat jednou a pak uz jen reinit, ne?

**Comment by dflanderka** [comment]
K druhé chybě:

FEValues::initialize se ve FieldFE::cache_update provádí pouze jednou při prvním použití, kdy se inicializují FEValues objekty pro všechny dimenze.

Do MultiField 'conc', což je ConcentrationTransportModel::ModelEqData::output_fields, se nastavuje DofHandler metodou set_fe_data, tím se nevolá FieldFE::set_mesh (viz. FieldFlags) a není vytvořen MixedPtr<FiniteElement> objekt. Ten by se mohl vzít s DofHandler.ds().fe(ElementAccessor)?

**Comment by jbrezmorf** [comment]
1. Try to introduce two constructors:
    a. FieldFE(mixed_ptr<FE>) creating its own dof handler, using approximately current make_dof_handler.
    b. FieldFE(dof_handler, vector,  component = 0) corresponding to current set_fe_data

2. Remove FieldFE::fe_, replace by a method fe() getting the FE from the dofhandler.

**Comment by jbrezmorf** [comment]
@jstebel FieldFE use FeValues for a fixed FE, we want to get this from the DiscreteSpace for the dof handler. However DiscreteSpace::fe(..) takes ElementAccessor as a parameter. We will temporarily modify this method, omitting the accessor parameter. However in order to allow more general discrete spaces we probably have to modify FEValues to be based on the DiscreateSpace instead of the FE. Or introduce another layer wrapping more FEValues for individual FE.

**Comment by jstebel** [comment]
> @jstebel FieldFE use FeValues for a fixed FE, we want to get this from the DiscreteSpace for the dof handler. However DiscreteSpace::fe(..) takes ElementAccessor as a parameter. We will temporarily modify this method, omitting the accessor parameter. However in order to allow more general discrete spaces we probably have to modify FEValues to be based on the DiscreateSpace instead of the FE. Or introduce another layer wrapping more FEValues for individual FE.

Temporarily I would suggest to define e.g. method DiscreteSpace::fe() without element accessor, which would work only if its type is EqualOrderDiscreteSpace. Otherwise, the general way would require a new "element FE handler" that would manage FEValues for all FE types within a given DiscreteSpace.

**Comment by jbrezmorf** [comment]
Changes to deal with Boundary integrals:
- use EdgePoint::point_on(side_accessor) in assembly instead of iterating over two ranges
- introduce PointBase, CoupllingPoint, and BoundaryPoint 
- CouplingPoint je jako EdgePoint + meotda lower_dim
- BoundaryPoint je jako BulkPoint, ale nema DHCellAccessor 
- introduce CouplingPoint::lower_dim() -> BulkPoint

**Comment by jbrezmorf** [comment]
Problem with zero dim eval points for BC of 1D.
Check where templated methods of `DimEvalPoints` are used, figure out how to allow dim 0 there.

**Comment by jbrezmorf** [comment]
Add TODO to remove virtual methods from Points classes, need changes also in `Field::operator()`.

**Comment by jbrezmorf** [comment]
Review interplay of 'FieldFE::set_fe_data' and 'FieldFE::set_mesh' the 'make_dof_handler' is called at the end of 'set_mesh' which may override dofhanlder set through 'set_fe_data'.

File: src/mesh/mesh.cc [hunk 1]
```diff
@@ -777,6 +777,45 @@ void Mesh::make_edge_permutations()
 			break;
 		}
 	}
+
+	for (vector<BoundaryData>::iterator bdr=boundary_.begin(); bdr!=boundary_.end(); bdr++)
+	{
+        if (bdr->bc_ele_idx_ >= element_vec_.size()) continue; // skip invalid boundary item
```
Comment (jstebel) [inline src/mesh/mesh.cc:783]: Shouldn't here be '<' in place of '>=' ?

File: src/mesh/mesh.cc [hunk 2]
```diff
@@ -777,6 +777,45 @@ void Mesh::make_edge_permutations()
...
+        Edge edg = this->edge(bdr->edge_idx_);
+        ElementAccessor<3> bdr_elm = this->element_accessor(bdr->bc_ele_idx_);
+
+        // node numbers is the local index of the node on the last side
+        // this maps the side nodes to the nodes of the side(0)
+        unsigned int n_side_nodes = bdr_elm->n_nodes();
+        unsigned int permutation[n_side_nodes];
+        node_numbers.clear();
```
Comment (jstebel) [inline src/mesh/mesh.cc:791]: Rather resize to n_side_nodes?

File: src/mesh/mesh.cc (hunk 1 above)
Comment (dflanderka) [reply src/mesh/mesh.cc:783]: See method Mesh::make_neighbours_and_edges. There are two main for cycles. We add boundary elements to mesh (not loaded from mesh file) in second of them. There is some 'hack' that adds two items to boundary_ vector for every new boundary element (but use only one item). We must skip unused (invalid) items in marked line.

**Pull request merged**